File: frontend/src/types/audit.ts
import type { DetectionResultItem, LocalDetectionResult } from '@/api/stream'
import type { PlanogramMatchResult } from '@/types/planogram'

/** Result returned by the shelf-image audit endpoint. */
export interface AuditAnalysisResult {
  /** Short recommended next step for store staff. */
  suggestedAction: string
  /** Longer reasoning that explains why the action was suggested. */
  explanation: string
  /** Persisted audit record id when the backend saved the run. */
  recordId?: string
  /** Closed-loop ticket ids created or updated by this audit. */
  ticketIds?: string[]
  /** Optional Detect → Decide → Dispatch narrative from the closed-loop agent. */
  closedLoopNarrative?: string | null
  /** JPEG data URL with model bounding boxes drawn on the shelf image. */
  annotatedImage?: string
  /** Structured local vision-model response sent to the agent layer. */
  visionModelResponse?: LocalDetectionResult
  detections?: DetectionResultItem[]
  detectionSummary?: LocalDetectionResult['summary']
  /** Active planogram match result (null when no planogram is selected). */
  planogramResponse?: PlanogramMatchResult | null
}

export type AuditRequestStatus = 'idle' | 'ready' | 'uploading' | 'analyzing' | 'success' | 'error'

export type AuditPipelineStep = 'vision' | 'planogram' | 'agent' | 'tickets' | 'done'

export type AuditStepStatus = 'pending' | 'running' | 'done' | 'skipped' | 'error'

export interface AuditStepState {
  id: AuditPipelineStep
  status: AuditStepStatus
}

export interface AuditPanelState {
  status: AuditRequestStatus
  previewUrl: string | null
  fileName: string | null
  result: AuditAnalysisResult | null
  errorMessage: string | null
  /** Progressive pipeline steps shown while analyzing. */
  steps: AuditStepState[]
  /** Currently active step id while analyzing. */
  activeStep: AuditPipelineStep | null
}

export const AUDIT_PIPELINE_STEPS: AuditPipelineStep[] = [
  'vision',
  'planogram',
  'agent',
  'tickets',
  'done',
]

export function createInitialAuditSteps(): AuditStepState[] {
  return AUDIT_PIPELINE_STEPS.map((id) => ({ id, status: 'pending' }))
}

File: backend/app/schemas/audit.py
"""Shelf-image audit schemas (mirrors ``frontend/src/types/audit.ts``)."""

from __future__ import annotations

from typing import Any

from app.schemas.common import CamelModel


class DetectionAgentRequest(CamelModel):
    """Vision-model JSON payload sent from the local detector to the agent."""

    vision_model_response: dict[str, Any]
    planogram_response: dict[str, Any] | None = None
    language: str = "en"
    # Optional client-side image for audit persistence (raw base64 or data URL).
    image_base64: str | None = None
    source_label: str | None = None


class AuditAnalysisResult(CamelModel):
    """Result returned by the shelf-image audit endpoint."""

    # Short recommended next step for store staff.
    suggested_action: str
    # Longer reasoning that explains why the action was suggested.
    explanation: str
    # Persisted record id when the audit was saved.
    record_id: str | None = None
    # Optional closed-loop ticket ids created/updated from this audit.
    ticket_ids: list[str] = []
    closed_loop_narrative: str | None = None
